src/libmongoc/CMakeLists.txt

```diff
@@ -1176,44 +1204,109 @@ set (VERSION "${MONGOC_VERSION}")
 set (prefix "${CMAKE_INSTALL_PREFIX}")
 set (libdir "\${prefix}/${CMAKE_INSTALL_LIBDIR}")
 
-foreach (
-      FLAG
-      ${SASL_LIBRARIES} ${SSL_LIBRARIES} ${SHM_LIBRARIES} ${RESOLV_LIBRARIES}
-      ${THREAD_LIB} ${ZLIB_LIBRARIES} ${SNAPPY_LIBRARIES} ${MONGOC_ZSTD_LIBRARIES} ${ICU_LIBRARIES}
+# Collect link items for the static library to be inserted into the pkg-config
+if(TARGET mongoc_static)
+   set(link_options
+      ${SASL_LIBRARIES} ${SSL_LIBRARIES} ${SHM_LIBRARIES} ${THREAD_LIB} ${ZLIB_LIBRARIES}
+      ${SNAPPY_LIBRARIES} ${MONGOC_ZSTD_LIBRARIES} ${ICU_LIBRARIES}
       ${LIBMONGOCRYPT_LIBRARY})
+   # Replace all absolute paths with search-dir link-file options:
+   list(TRANSFORM link_options
+        REPLACE "^(.+)/lib([^/]+)\\.[a-z]+$"
+                "-L\\1;-l\\2"
+        REGEX "^/")
+   # Prepend "-l" to all bare names:
+   list(TRANSFORM link_options PREPEND "-l" REGEX "^[^-]")
+   list(REMOVE_DUPLICATES link_options)
+   message(DEBUG "Computed static library link options: ${link_options}")
+   set_property(TARGET mongoc_static APPEND PROPERTY pc_LIBS ${link_options})
+endif()
 
-   if (IS_ABSOLUTE "${FLAG}")
-      get_filename_component (FLAG_DIR "${FLAG}" DIRECTORY)
-      get_filename_component (FLAG_FILE "${FLAG}" NAME_WE)
-      STRING (REGEX REPLACE "^lib" "" FLAG_FILE "${FLAG_FILE}")
-      set (MONGOC_LIBRARIES "${MONGOC_LIBRARIES} -L${FLAG_DIR} -l${FLAG_FILE}")
-   elseif (FLAG MATCHES "^-.*")
-      # Flag starts with dash, add it as-is.
-      set (MONGOC_LIBRARIES "${MONGOC_LIBRARIES} ${FLAG}")
-   else ()
-      # Flag doesn't start with dash, add it with a dash.
-      set (MONGOC_LIBRARIES "${MONGOC_LIBRARIES} -l${FLAG}")
-   endif ()
-endforeach ()
+# More pkg-config properties:
+set_target_properties(${TARGETS_TO_INSTALL} PROPERTIES
+   pc_NAME "libmongoc"
+   pc_DESCRIPTION "The libmongoc MongoDB client library."
+   pc_VERSION "${VERSION}")
+# Relative include-path will be given the install prefix:
+set_property(TARGET ${TARGETS_TO_INSTALL} APPEND PROPERTY pc_INCLUDE_DIRECTORIES "${MONGOC_HEADER_INSTALL_DIR}")
+
+#[[
+Utility for configuring a file and then evaluating generator expressions on it.
+The content of the file has a tiny templating DSL for writing generator
+expressions. Expansion takes place as follows:
+   1. Delete any content starting at hash-tilde "#~" until the next newline
+   2. Search for any lines of the form "%define foo bar". For each instance,
+      replace "foo" with "bar" in the remainder of the file. The "% foo bar"
+      line is deleted.
+   3. If a tilde "~" appears adjacent to any of "$,<:>@", delete the tilde and
+      all adjacent whitespace.
+   4. Run string(CONFIGURE ... @ONLY) on the result to replace variable
+      references. Special replacements @space@, @newline, and @empty@ are also
+      available for adjusting whitespace after ~ trimming.
+   5. Run file(GENERATE).
+]]
+function(_mongoc_configure_and_generate_file input output)
```

Comment:
Disregard the bit about lists vs. properties, the current target-based pattern is fine. The suggestion is just to move the logic into a module file instead of in config file generation.

Reply:
The DSL is quite a heavy hammer, but I went that route to utilize target properties, which are only definitely known once configuration completes and can use generator expressions to define their values, even in the face of multi-config generation. The "hope" is that one rarely needs to look inside the `.pc.in`.

(The `-targets.cmake` files generated by `install(TARGETS...EXPORT...)` do not use `@VAR@` substitution and are spit out by CMake directly. You are probably thinking of the mongoc-1.0-config.cmake file, which `includes()` the `-targets.cmake` file, which can be generated by arbitrary means, but often using `configure_package_config_file`.)